astyanax-recipes/src/main/java/com/netflix/astyanax/recipes/storage/ObjectReader.java

```diff
@@ -165,9 +181,15 @@ public void run() {
                         }
                     }
                     finally {
-                        executor.shutdown();
-                        if (!executor.awaitTermination(maxWaitTimeInSeconds, TimeUnit.SECONDS)) {
-                            throw new Exception("Took too long to fetch object: " + objectName);
+                        if (!isExecutorServiceProvided) {
+                            executor.shutdown();
+                            try {
+                                if (!executor.awaitTermination(maxWaitTimeInSeconds, TimeUnit.SECONDS))
+                                    throw new Exception("Took too long to fetch object: " + objectName);
+                            } finally {
+                                if (!executor.isTerminated())
+                                    executor.shutdownNow();
```

Comment:
+1 Logging is a good idea.

The only reliable way to kill executor threads still not terminated after shutdownNow() is to kill the JVM itself; in such case we'd need to further investigate whether this recipe is the root cause. If the block is due to an optionally-specified callback, the caller would need to investigate their callback logic.